src/ledger/LedgerState.cpp

```diff
@@ -0,0 +1,1252 @@
+        /*auto iter = mParent->mState.find(state.first);
+        if (iter == mParent->mState.end())
+        {
+            auto ler = makeStateEntry(entry, nullptr);
+            ler->invalidate();
+            mParent->mState[state.first] = ler;
+        }
+        else
+        {
+            auto const& previous =
+                iter->second->ignoreInvalid().previousEntry();
+            auto ler = makeStateEntry(entry, previous);
+            ler->invalidate();
+            iter->second = ler;
+        }*/
+    }
+}
+
+void
+LedgerState::mergeHeaderIntoParent()
+{
+    if (mHeader)
+    {
+        auto const& header = mHeader->ignoreInvalid().header();
+        auto const& previous = mHeader->ignoreInvalid().previousHeader();
+        auto lhr = makeStateHeader(header, previous);
+        lhr->invalidate();
+        mParent->mHeader = lhr;
+    }
+}
+
+void
+LedgerState::mergeStateIntoRoot()
+{
+    for (auto const& state : mState)
+    {
+        state.second->invalidate();
+        if (state.second->ignoreInvalid().entry())
+        {
+            storeInDatabase(state.second);
+        }
+        else
+        {
+            deleteFromDatabase(state.second);
+        }
+    }
+}
+
+void
+LedgerState::mergeHeaderIntoRoot()
+{
+    if (!mHeader)
+    {
+        loadHeader()->invalidate();
+    }
+    mRoot->setCurrentHeader(mHeader->ignoreInvalid().header());
+}
+
+void
+LedgerState::rollback()
+{
+    assert(mParent || mRoot);
+
+    if (mChild)
+    {
+        mChild->rollback();
+    }
+
+    if (mParent)
+    {
+        mParent->mChild = nullptr;
+        mParent = nullptr;
+    }
+    else
+    {
+        mRoot->hasChild(false);
+        mRoot = nullptr;
+        mTransaction->rollback();
+    }
+
+    mHeader.reset();
+    mState.clear();
+    mLoadBestOfferContext.reset();
+}
+
+LedgerState::StateEntry
+LedgerState::create(LedgerEntry const& entry)
+{
+    assert(mParent || mRoot);
+    assert(!mChild);
+
+    auto key = LedgerEntryKey(entry);
+    auto ler = createHelper(entry, key);
+    mState[key] = ler;
+
+    if (entry.data.type() == OFFER)
+    {
+        OfferEntry const& offer = entry.data.offer();
+        invalidateLoadBestOfferContext(offer.selling, offer.buying);
+    }
+
+    return ler;
+}
+
+LedgerState::StateEntry
+LedgerState::createHelper(LedgerEntry const& entry, LedgerKey const& key)
+{
+    StateEntry ler;
+
+    auto iter = mState.find(key);
+    if (iter != mState.end())
+    {
+        // This condition handles cases like the following example:
+        //    LedgerKey key = ...;
+        //    LedgerState lsParent(...);
+        //    LedgerState lsChild(lsParent);
+        //    lsChild.load(key)->erase();
+        //    lsChild.commit();
+        //    lsParent.create(key);
+        if (iter->second->ignoreInvalid().entry())
+        {
+            throw std::runtime_error("Key already exists in memory");
+        }
+
+        auto newEntry = std::make_shared<LedgerEntry>(entry);
+        if (!mChild)
```

Comment:
Please add comment to explain why this condition is needed and why when there is mChild, previous value is nullptr.